src/session.ts

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { IClientSession } from '@jupyterlab/apputils';
+
+import { CodeEditor } from '@jupyterlab/codeeditor';
+
+import { KernelMessage } from '@jupyterlab/services';
+
+import { PromiseDelegate } from '@phosphor/coreutils';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+import { DebugProtocol } from 'vscode-debugprotocol';
+
+import { IDebugger } from './tokens';
+
+export class DebugSession implements IDebugger.ISession {
+  /**
+   * Instantiate a new debug session
+   *
+   * @param options - The debug session instantiation options.
+   */
+  constructor(options: DebugSession.IOptions) {
+    this.client = options.client;
+    this.client.iopubMessage.connect(this._handleEvent, this);
+  }
+
+  /**
+   * Dispose the debug session.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this.client.iopubMessage.disconnect(this._handleEvent, this);
+    this._isDisposed = true;
+    this._disposed.emit();
+  }
+
+  /**
+   * A signal emitted when the debug session is disposed.
+   */
+  get disposed(): ISignal<this, void> {
+    return this._disposed;
+  }
+
+  /**
+   * Whether the debug session is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Start a new debug session
+   */
+  public async start(): Promise<void> {
+    await this.sendRequest('initialize', {
+      clientID: 'jupyterlab',
+      clientName: 'JupyterLab',
+      adapterID: 'python',
+      pathFormat: 'path',
+      linesStartAt1: true,
+      columnsStartAt1: true,
+      supportsVariableType: true,
+      supportsVariablePaging: true,
+      supportsRunInTerminalRequest: true,
+      locale: 'en-us'
+    });
+
+    await this.sendRequest('attach', {});
+  }
+
+  /**
+   * Stop the running debug session.
+   */
+  public async stop(): Promise<void> {
+    await this.sendRequest('disconnect', {
+      restart: false,
+      terminateDebuggee: true
+    });
+  }
+
+  /**
+   * Send a custom debug request to the kernel.
+   * @param command debug command.
+   * @param args arguments for the debug command.
+   */
+  async sendRequest<K extends keyof DebugSession.IDebugRequestTypes>(
+    command: K,
+    args: DebugSession.IDebugRequestTypes[K]
+  ): Promise<DebugSession.IDebugResponseTypes[K]> {
+    const message = await this._sendDebugMessage({
+      type: 'request',
+      seq: this._seq++,
+      command,
+      arguments: args
+    });
+    return message.content as DebugSession.IDebugResponseTypes[K];
+  }
+
+  /**
+   * Signal emitted for debug event messages.
+   */
+  get eventMessage(): ISignal<DebugSession, DebugProtocol.Event> {
```

Comment:
We can consider aliasing `DebugProtocol.Event` to `IDebugger.ISession.Event` here (generic event)

Reply:
👍